packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
+                processExit(ctx, exit, poolTx.signature);
+            }
+        }
+
+        // Deposit/Withdraw to/from the AMM account when necessary
+        for (uint i = 0; i < ctx.tokens.length; i++) {
+            if (ctx.ammBalances[i] > ctx.ammBalancesInAccount[i]) {
+                uint amount = ctx.ammBalances[i] - ctx.ammBalancesInAccount[i];
+                processDeposit(ctx, ctx.tokens[i], amount);
+            } else if (ctx.ammBalancesInAccount[i] > ctx.ammBalances[i]) {
+                uint amount = ctx.ammBalancesInAccount[i] - ctx.ammBalances[i];
+                processWithdrawal(ctx, ctx.tokens[i], amount);
+            }
+        }
+
+        // The ending AMM updates
+        processAmmUpdates(ctx, false);
+
+        emit QueueItemsProcessed(poolTransactions.length);
+        return ctx.numTransactionsConsumed;
+    }
+
+    function totalSupply()
+        public
+        view
+        returns (uint256)
+    {
+        return poolSupply;
+    }
+
+    function mint(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.add(amount);
+        poolBalance[owner] = poolBalance[owner].add(amount);
+    }
+
+    function burn(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.sub(amount);
+        poolBalance[owner] = poolBalance[owner].sub(amount);
+    }
+
+    function processAmmUpdates(
+        Context memory ctx,
+        bool   start
+        )
+        internal
+    {
+        for (uint i = 0; i < ctx.tokens.length; i++) {
             // Check that the AMM update in the block matches the expected update
-            AmmUpdateTransaction.AmmUpdate memory update = _block.readAmmUpdate(txIdx++);
+            AmmUpdateTransaction.AmmUpdate memory update = ctx._block.readAmmUpdate(ctx.txIdx++);
             require(update.owner == address(this), "INVALID_TX_DATA");
             require(update.accountID == accountID, "INVALID_TX_DATA");
-            require(update.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+            require(update.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
```

Comment:
Remind me this: we don't check the layer2 balances here because the circuit will guarantee the balances are indeed correct so the operator cannot lie about them?

Reply:
Correct, the operator has no control over the balance stored in the DA (if it puts in invalid data the block is invalid).